skypy/halo/mass.py

```diff
@@ -1,38 +1,319 @@
+    This function computes the halo mass function, defined
+    in equation 7.46 in [1]_.
+
+    Parameters
+    -----------
+    M : (nm,)
+        Array for the halo mass, in units of solar masses.
+    wavenumber : (nk,) array_like
+        Array of wavenumbers at which the power spectrum is evaluated,
+        in units of [Mpc^-1].
+    power_spectrum: (nk,) array_like
+        Linear power spectrum at redshift 0 in [Mpc^3].
+    redshift : float
+        Redshift value at which to evaluate the variance of the power spectrum.
+    cosmology : astropy.cosmology.Cosmology
+        Cosmology object providing methods for the evolution history of
+        omega_matter and omega_lambda with redshift.
+    collapse_function: function
+        Collapse function to choose from a variety of models:
+        `sheth_tormen_collapse_function`, `press_schechter_collapse_function`.
+    params: tuple
+        List of parameters that determines the model used for
+        the collapse function.
+
+    Returns
+    --------
+    mass_function: (nm,) array_like
+        Halo mass function for a given mass array, cosmology and redshift, in
+        units of :math:`Mpc^{-3} M_{Sun}^{-1}`.
+
+    Examples
+    ---------
+    >>> import numpy as np
+    >>> from skypy.halo import mass
+    >>> from skypy.power_spectrum import _eisenstein_hu as eh
+
+    This example will compute the halo mass function for elliptical and
+    spherical collapse, for a Planck15 cosmology.
+    The power spectrum is given by the Eisenstein and Hu fitting formula:
+
+    >>> from astropy.cosmology import Planck15
+    >>> cosmo = Planck15
+    >>> redshift = 0.0
+    >>> k = np.logspace(-3, 1, num=1000, base=10.0)
+    >>> A_s, n_s = 2.1982e-09, 0.969453
+    >>> Pk = eh.eisenstein_hu(k, A_s, n_s, cosmo, kwmap=0.02, wiggle=True)
+
+    The Sheth and Tormen mass function at redshift 0:
+
+    >>> m = 10**np.arange(9.0, 12.0, 2)
+    >>> mass.halo_mass_function(m, k, Pk, 0, cosmo,
+    ...     sheth_tormen_collapse_function, params=(0.3222, 0.707, 0.3, 1.686))
+    array([3.07523240e-11, 6.11387743e-13])
+
+    And the Press-Schechter mass function at redshift 0:
+
+    >>> mass.press_schechter_mass_function(m, k, Pk, 0, cosmo)
+    array([3.46908809e-11, 8.09874945e-13])
+
+    References
+    ----------
+    .. [1] Mo, H. and van den Bosch, F. and White, S. (2010), Cambridge
+        University Press, ISBN: 9780521857932.
+    '''
+    k = wavenumber
+    Pk = power_spectrum
+    z = redshift
+
+    sigma = np.sqrt(_sigma_squared(M, k, Pk, z, cosmology))
+    f_c = collapse_function(sigma, params)
+
+    dlognu_dlogm = _dlns_dlnM(sigma, M)
+    rho_bar = (cosmology.critical_density(0).to(u.Msun / u.Mpc ** 3)).value
+    rho_m0 = cosmology.Om(0) * rho_bar
+
+    return rho_m0 * f_c * dlognu_dlogm / np.power(M, 2)
+
+
+def halo_mass_sampler(m_min, m_max, resolution, wavenumber, power_spectrum,
+                      redshift, cosmology, mass_function, params, size=None):
```

Comment:
```suggestion
def halo_mass_sampler(m_min, m_max, resolution, wavenumber, power_spectrum,
                      redshift, cosmology, collapse_function, params, size=None):
```
Pass the collapse function here rather than the mass function and it will simplify the use of partial